src/Servers/Kestrel/Core/src/Internal/ServerAddressesCollection.cs

```diff
@@ -0,0 +1,195 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Diagnostics;
+using Microsoft.AspNetCore.Hosting.Server.Features;
+
+namespace Microsoft.AspNetCore.Server.Kestrel.Core.Internal
+{
+    internal class ServerAddressesCollection : ICollection<string>
+    {
+        private readonly List<string> _addresses = new List<string>();
+        private readonly PublicServerAddressesCollection _publicCollection;
+
+        public ServerAddressesCollection()
+        {
+            _publicCollection = new PublicServerAddressesCollection(this);
+        }
+
+        public ICollection<string> PublicCollection => _publicCollection;
+
+        public bool IsReadOnly => false;
+
+        public int Count
+        {
+            get
+            {
+                lock (_addresses)
+                {
+                    return _addresses.Count;
+                }
+            }
+        }
+
+        public void PreventPublicMutation()
+        {
+            lock (_addresses)
+            {
+                _publicCollection.IsReadOnly = true;
+            }
+        }
+
+        public void Add(string item)
+        {
+            lock (_addresses)
+            {
+                _addresses.Add(item);
+            }
+        }
+
+        public bool Remove(string item)
+        {
+            lock (_addresses)
+            {
+                return _addresses.Remove(item);
+            }
+        }
+
+        public void Clear()
+        {
+            lock (_addresses)
+            {
+                _addresses.Clear();
+            }
+        }
+
+        public void InternalAdd(string item)
```

Comment:
What's the point of InternalAdd/Remove/Clear? They do the same thing as the public ones.

Reply:
I thought I deleted those...